keps/sig-storage/20200217-namespace-transfer.md

```diff
@@ -0,0 +1,283 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP proposes an API to securely transfer Kubernetes storage resources between namespaces.
+The initial implementation will focus on PersistentVolumeClaims.  But the API could be applied
+to other resource types as compelling use cases come up.
+
+`StorageTransferRequest` resources are created in the target namespace.
+`StorageTransferApproval` resources are created in the source namespace.
+When a controller detects matching request/approval resources, a transfer in initiated.
+
+When transferring a PersistentVolumeClaim, the associated PersistentVolume will be updated,
+the source PersistentVolumeClaim will be deleted, and the target PersistentVolumeClaim will be created.
+The transfer simply deals with Kubernetes API resources.
+No data on the physical PersistentVolume is accessed/copied. 
+
+## Motivation
+
+Give Kubernetes users the flexibility to easily and securely share data between namespaces.
+
+### Goals
+
+- Define an API for transferring persistent storage resources between namespaces.
+- The API should be compatible with as many existing storage provisioners as possible.
+
+### Non-Goals
+
+- Transfer of resources other than PersistentVolumeClaims will not be discussed.
+- Implementation details only discussed in the context of influencing the API definition.
+
+## Proposal
+
+`StorageTransferRequest` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: transfer-foo
+  namespace: target-namespace
+spec:
+  source:
+    name: foo
+    namespace: source-namespace
+    kind: PersistentVolumeClaim
+  targetName: bar #this is optional
+```
+
+`StorageTransferApproval` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferApproval
+metadata:
+  name: approve-foo
+  namespace: source-namespace
+spec:
+  source:
+    name: foo
+    kind: PersistentVolumeClaim
+  targetNamespace: target-namespace
+```
+
+When matching `StorageTransferRequest` and `StorageTransferApproval` resources are detected, the transfer process is initiated.
+
+Matching requests/approvals is similar to the process for binding PersistentVolumes with PersistentVolumeClaims.
+If they are not explicitly bound by the user, a controller will attempt to pair them by looking for matching resource types/names.
+
+Once matching is complete, a controller will begin the transfer process which does the following (not necessarily in order):
+
+- Make sure the associated PersistentVolume is not deleted/recycled during the transfer by setting reclaim policy to `Retain` (if not already).
+- Delete the PVC `source-namespace\foo`.
+- Bind The PersistentVolume to `target-namespace\bar`.
```

Comment:
Earlier, targetName (with value of bar) is said to be optional.
If the value is omitted, how would the transfer work ?

Reply:
Personally I'd prefer not to be implicit, that's one of the reasons I proposed using the datasource model, we can easily enforce things like quota, limits and required claim parameters that way IMO.

Reply:
My main philosophy has been just to make it as close to the existing create workflow as possible and avoid assumptions.  It also means you don't have to duplicate parameters or testing for them.  Instead you use the existing create implementation as your foundation and build on it rather than trying to make it better or offering a divergent path.  Just my 2 cents though.